### PR TITLE
[MO, Frontends] Implement ChangePlaceholderType as Offline Transformation

### PR DESCRIPTION
**Pull request by rkazants**
**Details:** 
- Move ChangePlaceholderType transformation (to generate OldApiMap with legacy type) into offline transformation stage
- Make both legacy MO and ONNX frontend to use ChangePlaceholderType offline transformation
- Cover the transformation with functional tests.

**Tickets:** 65267

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [ ]  Unit tests;
* [x]  Framework operation tests N/A - no new FW operations enabled
* [x]  Transformation tests N/A - no new transformations implemented
* [ ]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list N/A
* [x]  Supported **public** models list N/A
* [x]  User guide update N/A

### CODE REVIEW

**Comment by rkazants** [comment]
@GlebKazantaev, could you please review? MO legacy transformation (for type conversion) is moved to common optimizations, wrapped with python bindings, and called from MO

**Review by popovaan — APPROVED** [review]

File: inference-engine/ie_bridges/python/src/openvino/offline_transformations/offline_transformations_api_impl_defs.pxd [hunk 1]
```diff
@@ -10,6 +10,8 @@ from ..inference_engine.ie_api_impl_defs cimport IENetwork
 cdef extern from "offline_transformations_api_impl.hpp" namespace "InferenceEnginePython":
     cdef void ApplyMOCTransformations(IENetwork network, bool cf)
 
+    cdef void ApplyMOCLegacyTransformations(IENetwork network)
```
Comment (GlebKazantaev) [inline inference-engine/ie_bridges/python/src/openvino/offline_transformations/offline_transformations_api_impl_defs.pxd:13]: Please move to the pybind11

File: inference-engine/src/transformations/src/transformations/common_optimizations/change_placeholder_types.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  return false;
+}
+}  // namespace
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::ChangePlaceholderTypes,
+                       "ChangePlaceholderTypes", 0);
+
+bool ngraph::pass::ChangePlaceholderTypes::run_on_function(
+    shared_ptr<Function> f) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/change_placeholder_types.cpp:35]: please fix code style

File: inference-engine/ie_bridges/python/src/openvino/offline_transformations/offline_transformations_api_impl_defs.pxd (hunk 1 above)
Comment (rkazants) [reply inference-engine/ie_bridges/python/src/openvino/offline_transformations/offline_transformations_api_impl_defs.pxd]: Done

**Comment by rkazants** [comment]
@tomdol, could you please take a look at this PR? I have added oofline transformation for OldApiMap with legacy type generation. ONNX frontend is also affected.

File: src/common/transformations/src/transformations/common_optimizations/change_placeholder_types.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+#include "ngraph/node.hpp"
+#include "ngraph/opsets/opset8.hpp"
+#include "transformations/rt_info/old_api_map_element_type_attribute.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+namespace {
+bool is_node_casts_to_float_or_shapeof(const Node* node) {
+    if (dynamic_cast<const opset8::ShapeOf*>(node)) {
```
[the remark this answers is not in the document]
Comment (tomdol) [reply src/common/transformations/src/transformations/common_optimizations/change_placeholder_types.cpp:20]: Perhaps it's better to check the op name stored in `DiscreteTypeInfo`?

**Review by tomdol — APPROVED** [review]
Overall LGTM, all added comments - for consideration.

File: src/common/transformations/include/transformations/common_optimizations/change_placeholder_types.hpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+#include <string>
+#include <vector>
+
+#include "openvino/pass/pass.hpp"
+#include "transformations_visibility.hpp"
+
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API ChangePlaceholderTypes;
```
[the remark this answers is not in the document]
Comment (rkazants) [reply src/common/transformations/include/transformations/common_optimizations/change_placeholder_types.hpp:16]: The idea is to have separate container with transformations for generation of OldApi map. So in future we potentially extend this container with one more transformation for legacy layout.

File: src/common/transformations/src/transformations/common_optimizations/change_placeholder_types.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rkazants) [reply src/common/transformations/src/transformations/common_optimizations/change_placeholder_types.cpp:20]: @tomdol, we use the same approach with explicit opsets in other transformations from MOCTransformations. Indeed, we will have to extend this pattern.

**Pull request merged**